cilium/CFP-34841-bgp-route-learning.md

```diff
@@ -0,0 +1,118 @@
+
+When configured for [Native-Routing](https://docs.cilium.io/en/stable/network/concepts/routing/#native-routing) mode, it appears possible to support the advanced network design described above.  However, doing so in large environments would require complex management and automation for Linux routing tables.
+
+Proposed here is a change to support route learning via BGP.
+
+## Motivation
+
+### User Stories
+
+![BGP Peering Diagram](./images/34841-bgp-route-learning.png)
+
+*As a cluster administrator, I wish to deploy Cilium on a node which connects to two or more routing domains.  The deployment is considered a success when Cilium is capable of learning non-overlapping routes using BGP from peers across each routing domain, and when Cilium is able to route outbound traffic to the correct routing domain.*
+
+
+https://github.com/cilium/cilium/pull/33035/
+
+
+## Goals
+
+* When Cilium is configured to use the Border Gateway Protocol (BGP) with one or more peers, routes advertised by those peers to Cilium are accepted, evaluated through BGP's Best Path Selection algorithm, and the resulting Best Path(s) installed into the node's dataplane.  The net result is that Cilium makes Internet Protocol (IP) routing decisions for Pod/Service sourced traffic based on routes learned via BGP.
+
+## Non-Goals
+
+* Making Cilium's BGP implementation Virtual Routing and Forwarding (VRF) instance aware.
+
+
+## Proposal
+
+### Overview
+
+### Route Programming
+
+Cilium supports two modes of routing -- [Encapsulation](https://docs.cilium.io/en/stable/network/concepts/routing/#encapsulation) and [Native](https://docs.cilium.io/en/stable/network/concepts/routing/#native-routing).
+
+#### Native Routing Mode
+
+When operating in Native Routing mode, the implementation appears straightforward.  A mechanism is required to synchronize routes learned from BGP, specifically Cilium's in-memory instance of GoBGP, to the Linux routing table.  For context, GoBGP may be deployed as a standalone daemon or instantiated directly through its Golang package.  Cilium instantiates directly.
+
+GoBGP on it's own is unable to install routes into the Linux routing table directly.  GoBGP's documentation [suggests](https://github.com/osrg/gobgp/blob/master/docs/sources/zebra.md) running an additional BGP daemon such as Quagga or FRR and establishing communication between the two.
+
+As the number of routes received, and rate of change may vary dramatically from one environment to another, it is the author's recommendation that one of the existing GoBGP supported-daemons be selected.  Under this model, an additional daemon will be deployed within Cilium's BGP-speaking pods.  For a Kubernetes-based deployment, these are the pods named `cilium-<hash>`.
```

Comment:
Cilium today supports multiple BGP instances. There are couple of options present, all Cilium BGP instances peer with single additional BGP daemon ( such as FRR, BIRD, or another gobgp+zebra ) which programs the learned routes into kernel. Or we create 1-1 additional BGP daemon and do peering over loopback IP + different port. 1-1 mapping sounds wrong in terms of scaling and potential conflicts while installing routes in kernel.

This seems bit counter intuitive 
- Additiionl BGP daemon -- Cilium BGP instances -- upstream routers.
- Alternatively, this might be better : cilium BGP instances -- Node BGP daemon (installs kernel routes)-- upstream routers

Few questions
1. With this design idea, why do we need to bundle additional BGP daemon into Cilium. To achieve similar result, we can have user controlled BGP daemon deployed on the node, which peers with upstream routers ( TORs/Core routers ). And Cilium peers with this BGP router on the node. Yes, there is additional complexity of managing this BGP daemon on the node, lifecycle of this daemon will have to be independent of kubernetes. 
2. GoBGP and zebra integration require additional testing, this maintenance burden will come to Cilium if we package them together and publish it as a bundle. How do we go about its maintenance.

Reply:
@harsimran-pabla This is great feedback!  Thank you for the depth of your responses.

For your first points, I agree with you.  I would not want to introduce N+ instances of a BGP daemon.  Your point about conflicts when programming the kernel routing table seems very likely.

**Adjacent BGP Daemon for RIB Programming**
When writing the proposal, I was modeling it after GoBGP's existing support -- a BGP daemon that is off to the side, but not inline.  Your point about graceful restart is important to consider here.  If the remote peer (the TOR) triggers a graceful restart, the downstream BGP daemon handling RIB programming would be unaware.  The GoBGP instance running via Cilium would remain running.   As long as Cilium's GoBGP instance remains running, the TOR's routes should remain in the adjacent BGP daemon.

**Inline BGP Daemon for RIB Programming**
This is an interesting idea.  Here, Cilium's BGP instance remains the furthest downstream from the network (a stub router).  Cilium peers with an intermediary BGP daemon on the node.  Under this model, would Cilium's BGP configuration reflect only the peering with the intermediary peer?

Reply:
> Under this model, would Cilium's BGP configuration reflect only the peering with the intermediary peer?

Yes, that would be the case. Cilium will only peer with on-node BGP instance, which would be listening on localhost and some specific port number.

> Do you have any perspective on how many users, and how large those Cilium installations are that have requested route learning in the past?

Not many, but few enough that we are considering this feature. This is mostly for advances topologies where multiple NICs are present on the server and there is some traffic-engineering going on.

> If I deploy and manage a separate BGP daemon, and it's now a dependency of Cilium, I would want to configure the additional daemon via CRDs as well. 

There are few cases where I have seen users installing additional BGP process on the node for other reasons. This comes from the need to provision the node itself prior to kubernetes installation and advertise node loopback address into core network via BGP. This is not possible with Cilium BGP, since it requires node to be part of kubernetes cluster already.

If you go towards the installing your own BGP router on the node, I would recommend looking at it from this angle as well. Decoupling kubernetes from this BGP process might provide some benefits.